lib/hotel_controller.rb

```diff
@@ -0,0 +1,86 @@
+module Hotel
+  class HotelController
+    attr_reader :rooms, :reservations, :blocks
+
+    def initialize
+      @rooms = []
+      @reservations = []
+      @blocks = []
+    end
+
+    def add_room(room)
+      @rooms << room
+    end
+
+    def list_all_rooms
+      return @rooms
+    end
+
+    def total_cost(reservation)
+      return reservation.total_cost
+    end
+
+    def find_available_rooms(given_date, leave_date = given_date)
+      if leave_date == given_date
+        unavailable_rooms = @reservations.select { |res| (res.arrive..res.depart).include?(given_date) }
+      else
+        unavailable_rooms = @reservations.select { |res| res.overlap?(given_date, leave_date) }
+      end
+      
+      unavailable_room_nums = unavailable_rooms.map{ |res| res.room_num }
+      available_rooms = @rooms.reject{ |room| unavailable_room_nums.include?(room.room_num)}
+      return available_rooms
+    end
+    
+    def reserve_rooms(arrive, depart, block = nil)
+      raise ArgumentError, "End date must be after start date" if depart < arrive || depart == arrive
```

Comment:
It's helpful to include the invalid arguments in the message of an `ArgumentError`:

```suggestion
      raise ArgumentError, "End date (#{depart}) must be after start date (#{arrive})" if depart < arrive || depart == arrive
```